pkg/resources/management.cattle.io/v3/project/validator.go

```diff
@@ -0,0 +1,211 @@
+	}
+	return admission.ResponseAllowed(), nil
+}
+
+func (a *admitter) admitCreateOrUpdate(oldProject, newProject *v3.Project) (*admissionv1.AdmissionResponse, error) {
+	projectQuota := newProject.Spec.ResourceQuota
+	nsQuota := newProject.Spec.NamespaceDefaultResourceQuota
+	if projectQuota == nil && nsQuota == nil {
+		return admission.ResponseAllowed(), nil
+	}
+	allowed, msg, err := checkQuotaFields(projectQuota, nsQuota)
+	if err != nil {
+		return nil, fmt.Errorf("error checking project fields: %w", err)
+	}
+	if !allowed {
+		return admission.ResponseBadRequest(msg), nil
+	}
+	err = a.checkQuotaValues(&nsQuota.Limit, &projectQuota.Limit, oldProject)
+	if err != nil {
+		return admission.ResponseBadRequest(err.Error()), nil
+	}
+	return admission.ResponseAllowed(), nil
+}
+
+func checkQuotaFields(projectQuota *v3.ProjectResourceQuota, nsQuota *v3.NamespaceResourceQuota) (bool, string, error) {
+	if projectQuota == nil && nsQuota != nil {
+		return false, "field resourceQuota is required when namespaceDefaultResourceQuota is set", nil
+	}
+	if projectQuota != nil && nsQuota == nil {
+		return false, "field namespaceDefaultResourceQuota is required when resourceQuota is set", nil
+	}
+
+	projectQuotaLimitMap, err := convert.EncodeToMap(projectQuota.Limit)
+	if err != nil {
+		return false, "", fmt.Errorf("failed to decode project quota limit: %w", err)
+	}
+	nsQuotaLimitMap, err := convert.EncodeToMap(nsQuota.Limit)
+	if err != nil {
+		return false, "", fmt.Errorf("failed to decode namespace default quota limit: %w", err)
+	}
+	if len(projectQuotaLimitMap) != len(nsQuotaLimitMap) {
+		return false, "resource quota and namespace default quota do not have the same resources defined", nil
+	}
+	for k := range projectQuotaLimitMap {
+		if _, ok := nsQuotaLimitMap[k]; !ok {
+			return false, fmt.Sprintf("missing namespace default for resource %s defined on resourceQuota", k), nil
+		}
+	}
+	return true, "", nil
+}
+
+func (a *admitter) checkQuotaValues(nsQuota, projectQuota *v3.ResourceQuotaLimit, oldProject *v3.Project) error {
+	// check quota on new project
+	err := namespaceQuotaFits(nsQuota, projectQuota)
+	if err != nil {
+		return err
+	}
+
+	// if there is no old project or no quota on the old project, no further validation needed
+	if oldProject == nil || oldProject.Spec.ResourceQuota == nil {
+		return nil
+	}
+
+	// check quota relative to used quota
+	err = usedQuotaFits(&oldProject.Spec.ResourceQuota.UsedLimit, projectQuota)
+	if err != nil {
+		return err
+	}
+
+	// check if default quota is enough to set on namespaces
+	namespaces, err := a.namespaceCache.List(labels.Set(map[string]string{"field.cattle.io/projectId": oldProject.Name}).AsSelector())
+	if err != nil {
+		return err
+	}
+	namespacesCount := len(namespaces)
+	var nsLimits []*v3.ResourceQuotaLimit
+	for i := 0; i < namespacesCount; i++ {
+		nsLimits = append(nsLimits, nsQuota)
+	}
+
+	return namespaceQuotaFitsForAll(nsLimits, projectQuota)
```

Comment:
From what I have read so far, this function is checking if the default limit, when applied to all namespaces in the project, would cause you to go over your limits. I'm not quite sure that this works as is though.

For example, lets say that I have 3 namespaces. My project limit is 9 total configmaps, and the current default is 2 configmaps, besides namespace 3 which has 4 configmaps (so total I'm using 8 out of my limit of 9). This check will say that since I have 9 total, 3 is an acceptable default. However, in reality, 3 will send me over my limit because of the non-default namespace. 

In short, do we need to take into account namespaces that are using over the default limit?

Reply:
Could you check https://github.com/rancher/rancher/blob/4479cfbde2f3e8bd0dc1363df8a8fb38f51bbdbe/pkg/api/norman/customization/project/project_store.go#L246-L276 to see if your concern applies to the existing implementation - the goal is parity with norman. We should open a separate issue to consider changing the behavior.

Reply:
Agree that we shouldn't change the behavior here. I'll resolve this for now in favor of separate comments on if/where the logic doesn't have parity with norman since that's the current goal.